File: site/src/preview.ts
// The preview, drawn live: the true arm at step k, and faded, the nominal
// model's rollout launched ghostSteps earlier from the then-true state
// under the torques actually issued. The layout matches the offline
// render in the Python media package, so the static image and the live
// canvas look the same. Autoplay only without a reduced-motion
// preference; stepping buttons always work.

import { motionOK } from "./anim/raf";
import { forwardKinematics } from "./arm/kinematics";
import { ARM, SCHEDULE } from "./arm/params";
import { simulateLoop, type Loop } from "./arm/schedule";
import { el } from "./dom";

const W = 960;
const H = 540;

interface Palette {
  paper: string;
  ink: string;
  inkSoft: string;
  line: string;
  steel: string;
  steelLight: string;
  stark: string;
  residual: string;
  residualDeep: string;
}

function readPalette(): Palette {
  const style = getComputedStyle(document.documentElement);
  const token = (name: string, fallback: string): string => {
    const value = style.getPropertyValue(name).trim();
    return value === "" ? fallback : value;
  };
  return {
    paper: token("--paper-raised", "#fbf9f3"),
    ink: token("--ink-strong", "#3e4a54"),
    inkSoft: token("--ink-soft", "#5f6c77"),
    line: token("--line", "#ddd3be"),
    steel: token("--steel", "#4d7b9e"),
    steelLight: token("--steel-light", "#a5c6df"),
    stark: token("--stark", "#eba538"),
    residual: token("--m-residual", "#7e57c2"),
    residualDeep: token("--m-residual-deep", "#6a46a8"),
  };
}

const FONT = 'system-ui, -apple-system, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif';

class Scene {
  private readonly base = { x: 0.63 * W, y: H - 0.31 * H };
  private readonly scale = 0.57 * H;
  private readonly strip = { left: 0.035 * W, right: 0.965 * W, top: H - 0.2 * H, bottom: H - 0.045 * H };
  private readonly limit: number;

  constructor(
    private readonly ctx: CanvasRenderingContext2D,
    private readonly loop: Loop,
    private readonly colors: Palette,
  ) {
    let limit = 1e-9;
    for (const r of loop.residual) limit = Math.max(limit, Math.abs(r[0]), Math.abs(r[1]));
    this.limit = limit;
  }

  private px(xy: readonly [number, number]): [number, number] {
    return [this.base.x + xy[0] * this.scale, this.base.y - xy[1] * this.scale];
  }

  private line(a: [number, number], b: [number, number], color: string, width: number, dash: number[] = []): void {
    const { ctx } = this;
    ctx.save();
    ctx.strokeStyle = color;
    ctx.lineWidth = width;
    ctx.lineCap = "round";
    ctx.setLineDash(dash);
    ctx.beginPath();
    ctx.moveTo(a[0], a[1]);
    ctx.lineTo(b[0], b[1]);
    ctx.stroke();
    ctx.restore();
  }

  private dot(c: [number, number], r: number, fill: string, stroke: string, width: number): void {
    const { ctx } = this;
    ctx.beginPath();
    ctx.arc(c[0], c[1], r, 0, 2 * Math.PI);
    ctx.fillStyle = fill;
    ctx.fill();
    ctx.lineWidth = width;
    ctx.strokeStyle = stroke;
    ctx.stroke();
  }

  private text(s: string, x: number, y: number, color: string, size: number, weight = "400", align: CanvasTextAlign = "left"): void {
    const { ctx } = this;
    ctx.fillStyle = color;
    ctx.font = `${weight} ${size}px ${FONT}`;
    ctx.textAlign = align;
    ctx.textBaseline = "middle";
    ctx.fillText(s, x, y);
  }

  draw(k: number): void {
    const { ctx, loop, colors } = this;
    const n = loop.states.length;
    const steps = SCHEDULE.ghostSteps;
    const launch = (((k - steps) % n) + n) % n;
    const ghost = loop.ghosts[launch] as readonly [number, number, number, number][];
    const ghostEnd = ghost[ghost.length - 1] as [number, number, number, number];
    const state = loop.states[k] as [number, number, number, number];

    ctx.fillStyle = colors.paper;
    ctx.fillRect(0, 0, W, H);

    // Reach arc and ground mark.
    const reach = (ARM.linkLengthsM[0] + ARM.linkLengthsM[1]) * this.scale;
    ctx.save();
    ctx.strokeStyle = colors.line;
    ctx.lineWidth = 1.2;
    ctx.setLineDash([4, 6]);
    ctx.beginPath();
    ctx.arc(this.base.x, this.base.y, reach, Math.PI, 2 * Math.PI);
    ctx.stroke();
    ctx.restore();
    this.line([this.base.x - 24, this.base.y], [this.base.x + 24, this.base.y], colors.line, 2.5);

    // The ghost: imagined hand path, faded arm, hollow hand.
    const ghostPose = forwardKinematics(ghostEnd[0], ghostEnd[1], ARM);
    const gElbow = this.px(ghostPose.elbow);
    const gHand = this.px(ghostPose.hand);
    ctx.save();
    ctx.strokeStyle = colors.steel;
    ctx.lineWidth = 1.8;
    ctx.setLineDash([3, 4]);
    ctx.beginPath();
    ghost.forEach((gs, i) => {
      const p = this.px(forwardKinematics(gs[0], gs[1], ARM).hand);
      if (i === 0) ctx.moveTo(p[0], p[1]);
      else ctx.lineTo(p[0], p[1]);
    });
    ctx.stroke();
    ctx.restore();
    this.line([this.base.x, this.base.y], gElbow, colors.steelLight, 11);
    this.line(gElbow, gHand, colors.steelLight, 11);
    this.dot(gHand, 9, colors.paper, colors.steel, 2.5);

    // The true arm and its recent hand trail.
    const pose = forwardKinematics(state[0], state[1], ARM);
    const elbow = this.px(pose.elbow);
    const hand = this.px(pose.hand);
    ctx.save();
    ctx.strokeStyle = colors.stark;
    ctx.globalAlpha = 0.8;
    ctx.lineWidth = 2.2;
    ctx.beginPath();
    for (let i = 0; i <= steps; i += 1) {
      const s = loop.states[(launch + i) % n] as [number, number, number, number];
      const p = this.px(forwardKinematics(s[0], s[1], ARM).hand);
      if (i === 0) ctx.moveTo(p[0], p[1]);
      else ctx.lineTo(p[0], p[1]);
    }
    ctx.stroke();
    ctx.restore();

    this.line(gHand, hand, colors.residual, 2.6);
    this.line([this.base.x, this.base.y], elbow, colors.ink, 11);
    this.line(elbow, hand, colors.ink, 11);
    this.dot([this.base.x, this.base.y], 8, colors.ink, colors.paper, 2);
    this.dot(elbow, 7, colors.ink, colors.paper, 2);
    this.dot(hand, 9, colors.stark, colors.ink, 2);

    const mid: [number, number] = [0.5 * (gHand[0] + hand[0]), 0.5 * (gHand[1] + hand[1])];
    const alongX = hand[0] - gHand[0];
    const alongY = hand[1] - gHand[1];
    const norm = Math.hypot(alongX, alongY) || 1;
    let perpX = -alongY / norm;
    let perpY = alongX / norm;
    if (perpY > 0) {
      perpX = -perpX;
      perpY = -perpY;
    }
    this.text("gap", mid[0] + 16 * perpX + 8, mid[1] + 16 * perpY - 4, colors.residual, 14, "700");

    // Labels.
    const x0 = 0.035 * W;
    const y0 = 0.075 * H;
    const aheadS = (steps * loop.dtS).toFixed(1);
    this.text("the model imagined", x0, y0, colors.steel, 19, "700");
    this.text(`nominal physics, ${aheadS} s ago`, x0, y0 + 30, colors.inkSoft, 14);
    this.text("the world did", x0, y0 + 80, colors.ink, 19, "700");
    this.text("payload, friction, actuator", x0, y0 + 110, colors.inkSoft, 14);
    this.text(`t = ${(k * loop.dtS).toFixed(1)} s`, W - 0.035 * W, y0, colors.inkSoft, 14, "400", "right");

    // Residual strip.
    const { left, right, top, bottom } = this.strip;
    const centre = 0.5 * (top + bottom);
    const half = 0.5 * (bottom - top) * 0.92;
    this.line([left, bottom], [right, bottom], colors.line, 1);
    this.line([left, centre], [right, centre], colors.line, 1, [2, 4]);
    this.text("the residual", left, top - 22, colors.residual, 19, "700");
    this.text(
      "acceleration the equations got wrong: shoulder (solid), elbow (dashed)",
      right,
      top - 22,
      colors.inkSoft,
      14,
      "400",
      "right",
    );
    const xs = (i: number): number => left + ((right - left) * i) / (n - 1);
    const ys = (v: number): number => centre - (half * v) / this.limit;
    const traces: { joint: 0 | 1; color: string; alpha: number; dash: number[] }[] = [
      { joint: 1, color: colors.residualDeep, alpha: 0.1, dash: [5, 3] },
      { joint: 0, color: colors.residual, alpha: 0.18, dash: [] },
    ];
    for (const { joint, color, alpha, dash } of traces) {
      ctx.save();
      ctx.beginPath();
      ctx.moveTo(xs(0), centre);
      for (let i = 0; i <= k; i += 1) ctx.lineTo(xs(i), ys((loop.residual[i] as [number, number])[joint]));
      ctx.lineTo(xs(k), centre);
      ctx.closePath();
      ctx.globalAlpha = alpha;
      ctx.fillStyle = color;
      ctx.fill();
      ctx.restore();
      ctx.save();
      ctx.strokeStyle = color;
      ctx.lineWidth = 2.2;
      ctx.setLineDash(dash);
      ctx.beginPath();
      for (let i = 0; i <= k; i += 1) {
        const p: [number, number] = [xs(i), ys((loop.residual[i] as [number, number])[joint])];
        if (i === 0) ctx.moveTo(p[0], p[1]);
        else ctx.lineTo(p[0], p[1]);
      }
      ctx.stroke();
      ctx.restore();
    }
    this.line([xs(k), top], [xs(k), bottom], colors.inkSoft, 1);
  }
}

export function mountPreview(host: HTMLElement): void {
  const loop = simulateLoop();
  const canvas = el("canvas", "preview-canvas");
  canvas.width = W;
  canvas.height = H;
  canvas.setAttribute("role", "img");
  canvas.setAttribute(
    "aria-label",
    "Two-link arm in its true world, with the nominal model's prediction from a moment " +
      "earlier drawn faded; the gap between the two hands is the model mismatch, and a " +
      "strip below traces the acceleration residual over time.",
  );
  const ctx = canvas.getContext("2d");
  if (ctx === null) throw new Error("canvas 2d context unavailable");
  const scene = new Scene(ctx, loop, readPalette());

  const n = loop.states.length;
  let frame = 0;
  let playing = false;
  let timer: number | null = null;

  const readout = el("p", "interactive-readout");
  const show = (k: number): void => {
    frame = ((k % n) + n) % n;
    scene.draw(frame);
    readout.textContent = `step ${frame + 1} of ${n}, t = ${(frame * loop.dtS).toFixed(2)} s`;
  };

  const controls = el("div", "control-row");
  const button = (label: string, onClick: () => void): HTMLButtonElement => {
    const b = el("button", "control-button", label);
    b.type = "button";
    b.addEventListener("click", onClick);
    controls.append(b);
    return b;
  };

  const stop = (): void => {
    if (timer !== null) window.clearInterval(timer);
    timer = null;
    playing = false;
    playButton.textContent = "play";
    playButton.setAttribute("aria-pressed", "false");
  };
  const start = (): void => {
    if (playing) return;
    playing = true;
    playButton.textContent = "pause";
    playButton.setAttribute("aria-pressed", "true");
    timer = window.setInterval(() => show(frame + 1), loop.dtS * 1000);
  };

  const playButton = button("play", () => (playing ? stop() : start()));
  button("previous", () => {
    stop();
    show(frame - 1);
  });
  button("next", () => {
    stop();
    show(frame + 1);
  });

  document.addEventListener("visibilitychange", () => {
    if (document.hidden) stop();
  });

  show(0);
  const caption = host.querySelector("figcaption");
  const stack = el("div", "interactive-stack");
  stack.append(canvas, controls, readout);
  host.querySelector("img")?.remove();
  if (caption !== null) host.insertBefore(stack, caption);
  else host.append(stack);
  if (motionOK()) start();
}

File: src/residual_worlds/media/preview.py
"""Render the animated preview: what the model imagined against what happened.

Frame k draws the true arm at step k and, faded, the nominal model's
rollout that was launched ``ghost_steps`` earlier from the then-true
state under the torques actually issued. The two hands land in different
places; that gap, per joint and in acceleration terms, is traced below
as the residual. Colors and type follow the site tokens so the image
sits naturally on the page.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt  # noqa: E402
import numpy as np  # noqa: E402
from matplotlib.backends.backend_agg import FigureCanvasAgg  # noqa: E402
from matplotlib.patches import Arc, Circle  # noqa: E402
from PIL import Image  # noqa: E402

from residual_worlds.config import ExperimentContract  # noqa: E402
from residual_worlds.media.loop import SCHEDULE, Loop, Schedule, simulate_loop  # noqa: E402
from residual_worlds.types import ArmParameters  # noqa: E402

PAPER = "#FBF9F3"
INK = "#3E4A54"
INK_SOFT = "#5F6C77"
LINE = "#DDD3BE"
STEEL = "#4D7B9E"
STEEL_LIGHT = "#A5C6DF"
STARK = "#EBA538"
RESIDUAL = "#7E57C2"
RESIDUAL_DEEP = "#6A46A8"

SIZE_LIMIT_BYTES = 3 * 1024 * 1024


@dataclass(frozen=True)
class RenderOptions:
    width: int = 960
    height: int = 540
    fps: int = 20
    webp_quality: int = 80
    gif_stride: int = 1  # write every n-th frame to the GIF to hold its size


def _fk(q: np.ndarray, arm: ArmParameters) -> tuple[np.ndarray, np.ndarray]:
    l1, l2 = arm.link_lengths_m
    elbow = np.array([l1 * np.cos(q[0]), l1 * np.sin(q[0])])
    hand = elbow + np.array([l2 * np.cos(q[0] + q[1]), l2 * np.sin(q[0] + q[1])])
    return elbow, hand


class _Frame:
    """One figure whose artists are updated per frame (no re-layout)."""

    def __init__(self, loop: Loop, arm: ArmParameters, schedule: Schedule, opt: RenderOptions):
        self.loop = loop
        self.arm = arm
        self.schedule = schedule
        self.opt = opt
        w, h = opt.width, opt.height
        self.fig = plt.figure(figsize=(w / 100, h / 100), dpi=100, facecolor=PAPER)
        self.canvas = FigureCanvasAgg(self.fig)

        # Arm panel: pixel coordinates, y up, base fixed.
        self.ax = self.fig.add_axes((0, 0, 1, 1))
        self.ax.set_xlim(0, w)
        self.ax.set_ylim(0, h)
        self.ax.set_axis_off()
        self.ax.set_facecolor(PAPER)
        self.scale = 0.57 * h
        self.base = np.array([0.63 * w, 0.31 * h])

        reach = sum(arm.link_lengths_m) * self.scale
        self.ax.add_patch(
            Arc(
                tuple(self.base),
                2 * reach,
                2 * reach,
                theta1=0,
                theta2=180,
                ec=LINE,
                lw=1.2,
                ls=(0, (4, 6)),
            )
        )
        self.ax.plot(
            [self.base[0] - 24, self.base[0] + 24],
            [self.base[1], self.base[1]],
            color=LINE,
            lw=2.5,
            solid_capstyle="round",
        )

        (self.ghost_path,) = self.ax.plot([], [], color=STEEL, lw=1.8, ls=(0, (3, 4)))
        (self.ghost_upper,) = self.ax.plot([], [], color=STEEL_LIGHT, lw=11, solid_capstyle="round")
        (self.ghost_lower,) = self.ax.plot([], [], color=STEEL_LIGHT, lw=11, solid_capstyle="round")
        self.ghost_hand = Circle((0, 0), 9, fill=True, fc=PAPER, ec=STEEL, lw=2.5)
        self.ax.add_patch(self.ghost_hand)

        (self.gap,) = self.ax.plot([], [], color=RESIDUAL, lw=2.6)
        (self.trail,) = self.ax.plot([], [], color=STARK, lw=2.2, alpha=0.8)
        (self.upper,) = self.ax.plot([], [], color=INK, lw=11, solid_capstyle="round")
        (self.lower,) = self.ax.plot([], [], color=INK, lw=11, solid_capstyle="round")
        self.joint_base = Circle(tuple(self.base), 8, fc=INK, ec=PAPER, lw=2)
        self.joint_elbow = Circle((0, 0), 7, fc=INK, ec=PAPER, lw=2)
        self.hand = Circle((0, 0), 9, fc=STARK, ec=INK, lw=2)
        for patch in (self.joint_base, self.joint_elbow, self.hand):
            self.ax.add_patch(patch)

        ahead_s = schedule.ghost_steps * loop.dt_s
        x0, y0 = 0.035 * w, h - 0.075 * h
        self.ax.text(
            x0, y0, "the model imagined", color=STEEL, fontsize=19, fontweight="bold", va="center"
        )
        self.ax.text(
            x0,
            y0 - 30,
            f"nominal physics, {ahead_s:.1f} s ago",
            color=INK_SOFT,
            fontsize=14,
            va="center",
        )
        self.ax.text(
            x0, y0 - 80, "the world did", color=INK, fontsize=19, fontweight="bold", va="center"
        )
        self.ax.text(
            x0, y0 - 110, "payload, friction, actuator", color=INK_SOFT, fontsize=14, va="center"
        )
        self.gap_label = self.ax.text(0, 0, "", color=RESIDUAL, fontsize=14, fontweight="bold")
        self.clock = self.ax.text(
            w - 0.035 * w, h - 0.075 * h, "", color=INK_SOFT, fontsize=14, ha="right", va="center"
        )

        # Residual strip along the bottom.
        left, right = 0.035 * w, 0.965 * w
        bottom, top = 0.045 * h, 0.20 * h
        self.strip = (left, right, bottom, top)
        self.ax.plot([left, right], [bottom, bottom], color=LINE, lw=1)
        self.ax.text(
            left,
            top + 10,
            "the residual",
            color=RESIDUAL,
            fontsize=19,
            fontweight="bold",
            va="bottom",
        )
        self.ax.text(
            right,
            top + 10,
            "acceleration the equations got wrong: shoulder (solid), elbow (dashed)",
            color=INK_SOFT,
            fontsize=14,
            ha="right",
            va="bottom",
        )
        self.limit = float(max(np.abs(loop.residual).max(), 1e-9))
        mid = 0.5 * (bottom + top)
        self.ax.plot([left, right], [mid, mid], color=LINE, lw=1, ls=(0, (2, 4)))
        (self.res_1,) = self.ax.plot([], [], color=RESIDUAL, lw=2.2)
        (self.res_2,) = self.ax.plot([], [], color=RESIDUAL_DEEP, lw=2.2, ls=(0, (5, 3)))
        (self.cursor,) = self.ax.plot([], [], color=INK_SOFT, lw=1)
        self.fills: list[Any] = []

    def _px(self, xy: np.ndarray) -> np.ndarray:
        return self.base + xy * self.scale

    def draw(self, k: int) -> Image.Image:
        loop, arm, sch = self.loop, self.arm, self.schedule
        n, h_steps = loop.frames, sch.ghost_steps
        launch = (k - h_steps) % n
        ghost = loop.ghosts[launch]

        g_elbow, g_hand = (self._px(p) for p in _fk(ghost[-1, :2], arm))
        elbow, hand = (self._px(p) for p in _fk(loop.states[k, :2], arm))

        ghost_hands = np.array([self._px(_fk(gs[:2], arm)[1]) for gs in ghost])
        self.ghost_path.set_data(ghost_hands[:, 0], ghost_hands[:, 1])
        self.ghost_upper.set_data([self.base[0], g_elbow[0]], [self.base[1], g_elbow[1]])
        self.ghost_lower.set_data([g_elbow[0], g_hand[0]], [g_elbow[1], g_hand[1]])
        self.ghost_hand.center = (g_hand[0], g_hand[1])

        idx = [(launch + i) % n for i in range(h_steps + 1)]
        trail = np.array([self._px(_fk(loop.states[i, :2], arm)[1]) for i in idx])
        self.trail.set_data(trail[:, 0], trail[:, 1])

        self.gap.set_data([g_hand[0], hand[0]], [g_hand[1], hand[1]])
        mid = 0.5 * (g_hand + hand)
        along = hand - g_hand
        norm = float(np.hypot(along[0], along[1])) or 1.0
        perp = np.array([-along[1], along[0]]) / norm
        if perp[1] < 0:
            perp = -perp
        self.gap_label.set_position((mid[0] + 16 * perp[0] + 8, mid[1] + 16 * perp[1] + 4))
        self.gap_label.set_text("gap")

        self.upper.set_data([self.base[0], elbow[0]], [self.base[1], elbow[1]])
        self.lower.set_data([elbow[0], hand[0]], [elbow[1], hand[1]])
        self.joint_elbow.center = (elbow[0], elbow[1])
        self.hand.center = (hand[0], hand[1])
        self.clock.set_text(f"t = {k * loop.dt_s:.1f} s")

        left, right, bottom, top = self.strip
        xs = left + (right - left) * np.arange(n) / (n - 1)
        centre = 0.5 * (bottom + top)
        half = 0.5 * (top - bottom) * 0.92
        upto = k + 1
        y1 = centre + half * loop.residual[:upto, 0] / self.limit
        y2 = centre + half * loop.residual[:upto, 1] / self.limit
        for fill in self.fills:
            fill.remove()
        self.fills = [
            self.ax.fill_between(xs[:upto], centre, y2, color=RESIDUAL_DEEP, alpha=0.10, lw=0),
            self.ax.fill_between(xs[:upto], centre, y1, color=RESIDUAL, alpha=0.18, lw=0),
        ]
        self.res_1.set_data(xs[:upto], y1)
        self.res_2.set_data(xs[:upto], y2)
        self.cursor.set_data([xs[k], xs[k]], [bottom, top])

        self.canvas.draw()  # type: ignore[no-untyped-call]
        rgba = np.asarray(self.canvas.buffer_rgba())  # type: ignore[no-untyped-call]
        return Image.fromarray(rgba[:, :, :3].copy(), "RGB")

    def close(self) -> None:
        plt.close(self.fig)


def render_frames(
    loop: Loop,
    arm: ArmParameters,
    schedule: Schedule = SCHEDULE,
    opt: RenderOptions | None = None,
) -> list[Image.Image]:
    frame = _Frame(loop, arm, schedule, opt or RenderOptions())
    try:
        return [frame.draw(k) for k in range(loop.frames)]
    finally:
        frame.close()


def write_animations(
    frames: list[Image.Image], output_dir: Path, opt: RenderOptions | None = None
) -> dict[str, Any]:
    opt = opt or RenderOptions()
    output_dir.mkdir(parents=True, exist_ok=True)
    duration_ms = int(round(1000 / opt.fps))
    webp = output_dir / "preview.webp"
    frames[0].save(
        webp,
        format="WEBP",
        save_all=True,
        append_images=frames[1:],
        duration=duration_ms,
        loop=0,
        quality=opt.webp_quality,
        method=6,
    )
    gif = output_dir / "preview.gif"
    gif_frames = frames[:: opt.gif_stride]
    gif_frames[0].save(
        gif,
        format="GIF",
        save_all=True,
        append_images=gif_frames[1:],
        duration=duration_ms * opt.gif_stride,
        loop=0,
        optimize=True,
    )
    sizes = {"webp_bytes": webp.stat().st_size, "gif_bytes": gif.stat().st_size}
    for name, size in sizes.items():
        if size > SIZE_LIMIT_BYTES:
            raise RuntimeError(f"{name} is {size} bytes, above the {SIZE_LIMIT_BYTES} byte limit")
    return {"webp": str(webp), "gif": str(gif), "frames": len(frames), **sizes}


def render_preview(
    contract: ExperimentContract,
    output_dir: Path,
    world_id: str = "composite_standard",
    opt: RenderOptions | None = None,
) -> dict[str, Any]:
    opt = opt or RenderOptions()
    loop = simulate_loop(contract, world_id, SCHEDULE)
    frames = render_frames(loop, contract.arm, SCHEDULE, opt)
    return {"world_id": world_id, **write_animations(frames, output_dir, opt)}
